README.md
# cerise-client
[![Develop build status](https://api.travis-ci.org/MD-Studio/cerise-client.svg?branch=develop)](https://travis-ci.org/MD-Studio/cerise-client) [![Documentation Status](https://readthedocs.org/projects/cerise_client/badge/?version=latest)](http://cerise_client.readthedocs.io/en/latest/?badge=latest) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/b75c0e30dc5340d4b5e1a5797c83fa57)](https://www.codacy.com/app/LourensVeen/cerise-client?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=MD-Studio/cerise-client&amp;utm_campaign=Badge_Grade) [![Codacy Badge](https://api.codacy.com/project/badge/Coverage/b75c0e30dc5340d4b5e1a5797c83fa57)](https://www.codacy.com/app/LourensVeen/cerise-client?utm_source=github.com&utm_medium=referral&utm_content=MD-Studio/cerise-client&utm_campaign=Badge_Coverage)[![DOI](https://zenodo.org/badge/102342552.svg)](https://zenodo.org/badge/latestdoi/102342552)

A simple Python client library for the Cerise job running service


CHANGELOG.rst
###########
Change Log
###########

All notable changes to this project will be documented in this file.
This project adheres to `Semantic Versioning <http://semver.org/>`_.

0.3.1
*****

* No-change re-release with Zenodo integration and DOI

0.3.0
*****

Fixed
-----

* Small bug fixes
* Various code quality issues
* API improvements

Added
-----

* More flexible file uploads

Removed
-------

* Docker container management (now Cerise Manager)
* Python 2.7 support


0.2.4
*****

Fixed
-----

* Several Docker handling issues


0.2.3
*****

Added
-----

* Arrays of input files
* Tell container which port it's running on


0.2.2
*****

Fixed
-----

* Better documentation
* Disallow running a job twice
* Small fixes to test suite
* Fixed Python 2 suppport
* Removed obsolete delete function

Added
-----

* SecondaryFile support
* Improved error handling
* Service.get_log()


0.2.1
*****

Fixed
-----

* Example in documentation
* Small fixes


0.2.0
*****

Fixed
-----

* Reworked API

Added
-----

* PyPI packaging configuration


0.1.0
*****

* Initial functionality


cerise_client/test/clean_up.py
import docker

def clean_up_service(srv_name):
    dc = docker.from_env()
    try:
        test_srv = dc.containers.get(srv_name)
        test_srv.stop()
        test_srv.remove()
    except docker.errors.NotFound:
        pass

# Clean up any mess left over from previous failed tests.
def clean_up():
    clean_up_service('cerise_client_test_service')
    clean_up_service('cerise_client_test_service2')




cerise_client/test/conftest.py
import docker
from pathlib import Path
import pytest
import requests
import time

from cerise_client import OutputFile, Service


def clean_up_service(srv_name):
    dc = docker.from_env()
    try:
        test_srv = dc.containers.get(srv_name)
        test_srv.stop()
        test_srv.remove()
    except docker.errors.NotFound:
        pass


# Clean up any mess left over from previous failed tests.
@pytest.fixture(scope="session")
def clean_up():
    clean_up_service('cerise_client_test_service')
    clean_up_service('cerise_client_test_service2')


@pytest.fixture(scope="session")
def docker_client(request):
    client = docker.from_env()
    yield client
    client.close()


@pytest.fixture(scope="session")
def test_image(clean_up, docker_client):
    """Get a plain cerise image for testing.

    Ignores errors; we may have a local image available already,
    in which case we want to continue, otherwise the other
    tests will fail.
    """
    try:
        docker_client.images.pull('mdstudio/cerise:develop')
    except docker.errors.APIError:
        pass
    return 'mdstudio/cerise:develop'


@pytest.fixture(scope="session")
def test_service(test_image, docker_client):
    env = {'CERISE_STORE_LOCATION_CLIENT': 'http://localhost:29593/files'}
    docker_client.containers.run(
            test_image, name='cerise_client_test_service',
            ports={'29593/tcp': ('127.0.0.1', 29593) },
            environment=env, detach=True)
    time.sleep(1)
    srv = Service('http://localhost', 29593)
    yield srv
    clean_up_service('cerise_client_test_service')


@pytest.fixture()
def this_dir():
    return Path(__file__).parent


@pytest.fixture()
def test_container(test_image, docker_client):
    try:
        container = docker_client.containers.get('cerise_client_test_service')
    except docker.errors.NotFound:
        image = docker_client.images.get(test_image)

        container = docker_client.containers.run(
                image,
                name='cerise_client_test_service',
                ports={'29593/tcp': ('127.0.0.1', 29593) },
                detach=True)

    yield container

    container.stop()
    container.remove()


@pytest.fixture()
def test_service_dict():
    return {
            'name': 'cerise_client_test_service',
            'port': 29593
            }


def create_test_job(test_service, this_dir, name):
    job = test_service.create_job(name)
    r = requests.get('http://localhost:29593/files/input/' + name + '/')
    assert r.status_code == 200

    # run job to create some outputs
    job.set_workflow(str(this_dir / 'test_workflow2.cwl'))
    job.add_input_file('input_file', str(this_dir / 'test_workflow2.cwl'))
    job.run()
    while job.is_running():
        time.sleep(0.1)
    assert job.state == 'Success'

    counts = job.outputs['counts']
    assert counts.text != ''
    return job


@pytest.fixture()
def test_file(request, test_service, this_dir):
    requests.put(
            'http://localhost:29593/files/input/test_workflow.cwl',
            data=(this_dir / 'test_workflow.cwl').open('rb'))
    return 'http://localhost:29593/files/input/test_workflow.cwl'

@pytest.fixture()
def test_output_file(request, test_file):
    return OutputFile(test_file)



cerise_client/test/fixtures.py
import docker
import os
from pathlib import Path
import pytest
import requests
import time

import cerise_client.service as cs

@pytest.fixture()
def docker_client(request):
    return docker.from_env()

@pytest.fixture(scope="session")
def test_image(request):
    """Get a plain cerise image for testing.

    Ignores errors; we may have a local image available already,
    in which case we want to continue, otherwise the other
    tests will fail.
    """
    docker_client = docker.from_env()
    try:
        docker_client.images.pull('mdstudio/cerise:develop')
    except docker.errors.APIError:
        pass
    return 'mdstudio/cerise:develop'

@pytest.fixture(scope="session")
def test_service(request, test_image):
    from .clean_up import clean_up_service
    clean_up_service('cerise_client_test_service')

    docker_client = docker.from_env()

    env = {'CERISE_STORE_LOCATION_CLIENT': 'http://localhost:29593/files'}
    docker_client.containers.run(
            test_image, name='cerise_client_test_service',
            ports={'29593/tcp': ('127.0.0.1', 29593) },
            environment=env, detach=True)
    time.sleep(1)
    srv = cs.Service('http://localhost', 29593)
    yield srv
    clean_up_service('cerise_client_test_service')

@pytest.fixture()
def this_dir():
    return Path(__file__).parent

def create_test_job(test_service, this_dir, name):
    job = test_service.create_job(name)
    r = requests.get('http://localhost:29593/files/input/' + name + '/')
    assert r.status_code == 200

    # run job to create some outputs
    job.set_workflow(str(this_dir / 'test_workflow2.cwl'))
    job.add_input_file('input_file', str(this_dir / 'test_workflow2.cwl'))
    job.run()
    while job.is_running():
        time.sleep(0.1)
    assert job.state == 'Success'

    counts = job.outputs['counts']
    assert counts.text != ''
    return job



examples/example.py
import cerise_client.service as cc
import time

srv = cc.require_managed_service(
            'my-cerise-service', 29593,
            'mdstudio/cerise:develop')
cc.start_managed_service(srv)

job = srv.create_job('example_job')
job.set_workflow('./example_workflow.cwl')
job.add_input_file('input_file', __file__)
job.run()

# Give the service a chance to stage the job
while job.state == 'Waiting':
    time.sleep(1)

persisted_srv = cc.service_to_dict(srv)
persisted_job = job.id

cc.stop_managed_service(srv)


# store the dict somewhere in persistent storage
# stop client software, shut down computer, etc.

# start computer and client software again
# get dict from persistent storage

srv = cc.service_from_dict(persisted_srv)
cc.start_managed_service(srv)
job = srv.get_job_by_id(persisted_job)

while job.is_running():
    time.sleep(10)

if job.state == 'Success':
    job.outputs['counts'].save_as('counts.txt')
else:
    print('There was an error: ' + job.state)
    print('Job log:')
    print(job.log)

srv.destroy_job(job)

cc.destroy_managed_service(srv)



cerise_client/job.py
from cerise_client import JobAlreadyExists, NoPrimaryFile, OutputFile

import os


class Job:
    def __init__(self, service, name, job_id=None, inputs=None,
            workflow=None, input_desc=None):
        """
        Create a new Job object.

        This will not actually make a job on the service. Use
        Service.create_job() to make a job.

        Args:
            service (Service): The service this job will run on.
            name (str): The name for the job.
        """
        self.id = job_id
        """str: The service-assigned id of this job."""

        self.name = name
        """str: The name of this job."""

        self._service = service
        """Service: The service that this job runs on."""

        self._inputs = inputs
        """[str]: List of declared inputs of the workflow."""
        if self._inputs is None: self._inputs = []

        self._workflow_url = workflow
        """str: The remote path to the uploaded workflow file."""

        self._input_desc = input_desc
        """dict: The input description object to be submitted."""
        if self._input_desc is None: self._input_desc = {}

        self._outputs = None
        """dict: Cached results."""

    @property
    def state(self):
        """
        The state that this job is in.

        None if the job has not been started yet. One of 'Waiting',
        'Running', 'Success', 'Cancelled', 'TemporaryFailure',
        'PermanentFailure', or 'SystemError'.
        """
        if self.id is None:
            return None
        return self._service._job_state(self.id)

    @property
    def log(self):
        """
        str: The job's log, as produced by the service.
        """
        if self.id is not None:
            return self._service._get_job_log(self.id)
        return None

    def set_workflow(self, workflow):
        """
        Sets the workflow file.

        Only one workflow file may be submitted; if this function is
        called repeatedly, the last workflow file is used.

        Valid types for workflow are:

        str: A path to a local CWL file containing the workflow.
        bytes: A bytes object containing a CWL workflow definition.
        File-like object: A stream to read the workflow content from.

        If you have the workflow content in string format, use
        set_workflow(wf.encode('utf-8')) (i.e. convert to bytes first).

        Args:
            workflow: Either a str containing the path to a file to
                    use, or a File-like object to read content from,
                    or a bytes object containing the workflow contents.
        """
        # scan workflow for inputs and store them in the object
        # list of (name, type) tuples? or {name: type}?
        # dict is better, need to search by name when adding

        # get file object from input
        if not isinstance(workflow, str):
            workflow = 'workflow.cwl', workflow

        # upload the workflow
        self._workflow_url = self._service._upload_file(self.name, workflow)

        # maybe make subdirs for input files and workflow, to allow
        # same names?

    def add_input_file(self, input_name, file_desc):
        """
        Adds an input file for the given workflow input.

        The set_workflow() function must be called before this one.
        The input name must correspond to an input defined in the
        workflow. The file's name must not equal that of any other
        file added using this function or add_secondary_file(), or
        `workflow.cwl`.

        The file_desc argument may be a string, a tuple of a string and
        a file-like object, a tuple of a string and a bytes object,
        or a list containing any of the above.

        If file_desc is a string, it is assumed to contain a path, and
        the corresponding file is uploaded and added as an input file
        with its original name. If file_desc is a tuple of a string and
        a file-like object, the string is used as the file name, and
        the content is read from the object. If file_desc is a tuple
        of a string and a bytes object, the bytes object is used as the
        content and the string as the name. If a list is supplied, each
        list item is processed as above, and the input's value is set
        to a corresponding array of Files.

        If this function is called repeatedly with the same input
        name, the last file(s) will be used.

        Note that this function will upload the input file(s) to the
        service for this job. If the file is large and/or the
        connection slow, then this will take a while.

        Args:
            input_name (str): The name of a workflow input.
            file_desc: A description of the file(s) to add, see above.

        Raises:
            UnknownInput: The input name does not match any in this
                workflow, or the workflow was not yet set.
            FileNotFound: A file to be used was not found.
        """
        # TODO: check against inputs
        if not isinstance(file_desc, list):
            file_descs = [file_desc]
        else:
            file_descs = file_desc

        input_descs = []
        for desc in file_descs:
            remote_url = self._service._upload_file(self.name, desc)
            if isinstance(desc, str):
                basename = os.path.basename(desc)
            elif isinstance(desc, tuple):
                basename = desc[0]
            input_descs.append({
                "class": "File",
                "location": remote_url,
                "basename": basename
                })

        if isinstance(file_desc, list):
            self._input_desc[input_name] = input_descs
        else:
            self._input_desc[input_name] = input_descs[0]

    def add_secondary_file(self, input_name, file_desc):
        """
        Adds a secondary file for the given workflow input.

        A primary file must be set using add_input_file() first.

        The input name must correspond to an input defined in the
        workflow. The file's name must not equal that of any other
        file added using this function or add_secondary_file(), or
        that of the workflow itself.

        The file_desc argument may be a string, a tuple of a string and
        a file-like object, a tuple of a string and a bytes object,
        or a list containing any of the above.

        If file_desc is a string, it is assumed to contain a path, and
        the corresponding file is uploaded and added as a secondary file
        with its original name. If file_desc is a tuple of a string and
        a file-like object, the string is used as the file name, and
        the content is read from the object. If file_desc is a tuple
        of a string and a bytes object, the bytes object is used as the
        content and the string as the name.

        Note that this function will upload the input file to the
        service for this job. If the file is large and/or the
        connection slow, then this will take a while.

        Args:
            input_name (str): The name of a workflow input.
            file_desc: The file to set, see above.

        Raises:
            UnknownInput: The input name does not match any in this
                workflow, or the workflow was not yet set.
            FileNotFound: The file to be used was not found.
            NoPrimaryFile: No primary file was set yet via \
                add_input_file().
        """
        if not input_name in self._input_desc:
            raise NoPrimaryFile('Primary file not set yet')

        remote_url = self._service._upload_file(self.name, file_desc)

        if isinstance(file_desc, str):
            basename = os.path.basename(file_desc)
        elif isinstance(file_desc, tuple):
            basename = file_desc[0]

        if not 'secondaryFiles' in self._input_desc[input_name]:
            self._input_desc[input_name]['secondaryFiles'] = list()

        self._input_desc[input_name]['secondaryFiles'].append({
                "class": "File",
                "location": remote_url,
                "basename": basename
                })

    def set_input(self, input_name, value):
        """
        Sets the value of a workflow input.

        Use this function for setting the value of numerical or string
        inputs. For File inputs, use add_input_file.

        Args:
            input_name (str): The name of a workflow input.
            value (Union[str,int,double]): The value to set.

        Raises:
            UnknownInput: The input name does not match any in this
                workflow, or the workflow was not yet set.
        """
        # TODO: check against inputs
        self._input_desc[input_name] = value

    def run(self):
        """
        Starts the job running on the associated service.

        Returns:
            str: The id given to this job by the service.
        """
        if self.id is not None:
            raise JobAlreadyExists()

        job_desc = {
                'name': self.name,
                'workflow': self._workflow_url,
                'input': self._input_desc
                }
        self.id = self._service._run_job(job_desc)
        return self.id

    def is_running(self):
        """
        Returns True when the job is still running, False when it is
        done.

        Returns:
            boolean: Whether the job is running.
        """
        state = self.state      # fetch only once
        return state == 'Waiting' or state == 'Running'

    def cancel(self):
        """
        Cancel this job; stop it running at the compute service.
        After this function is called, the job's state will eventually
        become 'Cancelled', unless it was already complete, in which
        case its state will be its normal final state.
        """
        if self.id is not None:
            self._service._cancel_job(self.id)

    @property
    def outputs(self):
        """
        Returns a dictionary of output objects. Keys are taken from
        the names of the outputs in the submitted workflows, the
        values are the corresponding results.

        If an output is of type File, an object of class
        OutputFile is returned as the value.

        If no outputs are available, returns None.

        Returns:
            Union[dict, None]: Output values or None.
        """
        if self._outputs is None:
            outputs = self._service._get_outputs(self.id)
            if outputs != {}:
                self._outputs = {}
                for key, value in outputs.items():
                    if isinstance(value, dict):
                        if value.get('class', '') == 'File':
                            self._outputs[key] = OutputFile(value['location'])
                    else:
                        self._outputs[key] = value

        return self._outputs


cerise_client/errors.py
class JobNotFound(Exception):
    """
    The given job does not exist on this service. Either it never
    existed because it hasn't been submitted yet (did you call
    job.run()?), or it was deleted.
    """
    pass

class JobAlreadyExists(Exception):
    """
    You tried to create a job with a name that already existed on this
    service. Jobs must have a unique name, and cannot be submitted
    twice.
    """
    pass

class InvalidJob(Exception):
    """
    You submitted an invalid job. Did you forget to add a workflow?
    """
    pass

class UnknownInput(Exception):
    """
    You tried to set the value for an input that is not in the workflow
    for the job you are creating. Did you add the workflow first?
    """
    pass

class NoPrimaryFile(Exception):
    """
    You tried to set a secondary file for an input for which no primary
    file is set. Use add_input_file() first, then add_secondary_file()
    with the same input name.
    """

class MissingOutput(Exception):
    """
    The output returned by the service does not refer to an existing
    file. Maybe the job was deleted?
    """

class FileNotFound(Exception):
    """
    The file you tried to set as a workflow input was not found. Is the
    path correct? Does the file exist?
    """

class CommunicationError(Exception):
    """
    There was an error communicating with the service.
    """
    pass


cerise_client/output_file.py
from cerise_client import MissingOutput

import requests


class OutputFile:
    def __init__(self, uri):
        """
        Creates a new OutputFile object.

        Args:
            uri (str): The URI at which the file is available.
        """
        self._uri = uri
        """The URI at which the file is available."""

    def save_as(self, file_path):
        """
        Downloads the file and saves it to disk.

        Args:
            file_path (str): The path to save the file to.
        Raises:
            IOError: There was a problem saving the file.
            MissingOutput: The output doesn't exist. Maybe the
                job was deleted?
        """
        with open(file_path, 'wb') as f:
            f.write(self._get_file().content)

    @property
    def text(self):
        """
        Returns the file in text form.

        Autodetects the encoding and converts to a standard Python
        unicode str.

        Returns:
            str: The contents of the file as text.
        Raises:
            MissingOutput: The output doesn't exist. Maybe the
                job was deleted?
        """
        return self._get_file().text

    @property
    def content(self):
        """
        Returns the file in binary form.

        Returns:
            bytes: The contents of the file as raw bytes.
        """
        return self._get_file().content

    def _get_file(self):
        """
        Returns a requests.Response object with the remote file.

        Returns:
            (requests.Response): The remote file
        Raises:
            MissingOutput: The output doesn't exist. Maybe the
                job was deleted?
        """
        r = requests.get(self._uri)
        if r.status_code == 404:
            raise MissingOutput()
        return r


cerise_client/service.py
from cerise_client import (CommunicationError, FileNotFound, InvalidJob, Job,
                           JobAlreadyExists, JobNotFound)

import docker

import errno
from io import BytesIO
import os
import requests
import tarfile
import tempfile
import time


class Service:
    def __init__(self, host='localhost', port=29593):
        """
        Create a new Service object representing an actual service.

        Note that this does not connect to the service, that is done
        as needed when the methods are called. So it will not raise
        even if the given service does not exist or you have no
        network connection.

        Args:
            host (str): The hostname of the service we are connecting
                to. May be prepended by 'http://' or 'https://', with
                https the default if a bare hostname is given.
            port (int): The port number on which the service runs.
        """
        if not host.startswith('http://') and not host.startswith('https://'):
            host = 'https://{}'.format(host)
        self._srv_loc = '{}:{}'.format(host, port)
        self._filestore = self._srv_loc + '/files'
        self._jobs = self._srv_loc + '/jobs'

    def create_job(self, job_name):
        """
        Creates a job to run on the given service.

        Args:
            job_name (str): The name of the job. This must be unique
                among all jobs on this service, and must consist of
                letters, digits, underscores and hyphens. A GUID or
                similar is recommended.

        Returns:
            Job: The new job.

        Raises:
            JobAlreadyExists: A job with this name already exists
                on this service.
        """
        self._create_input_dir(job_name)
        return Job(self, job_name)

    def destroy_job(self, job):
        """
        Removes a job, including all its input and output data,
        from the service.

        Args:
            job (Job): The job to be removed.

        Raises:
            JobNotFound: No job with this name or id was found on this
                service. Did you destroy it twice?
            CommunicationError: There was an error communicating with
                the service.
        """
        self._delete_input_dir(job.name)
        r = requests.delete(self._jobs + '/' + job.id)

        if r.status_code == 404:
            raise JobNotFound("The job could not be found.")

        if r.status_code != 204:
            raise CommunicationError(r)


    def get_job_by_id(self, job_id):
        """
        Gets a job by its id.

        Args:
            job_id (str): The id of the job, as returned by job.id
                after it's been created.

        Returns:
            Job: The requested job.

        Raises:
            JobNotFound: The job is unknown to this service. Either
                it was deleted, or it never existed on this service.
        """
        job = self._get_job_from_service(job_id)
        declared_inputs = None
        return Job(self, job['name'], job['id'],
                declared_inputs, job['workflow'], job['input'])

    def get_job_by_name(self, job_name):
        """
        Gets a job by its name. Only jobs that have been submitted
        are on the service, so if job.run() has not been called,
        this will not find the job.

        Args:
            job_name (str): The name of the job, as set when it was
                created.

        Returns:
            Job: The requested job.

        Raises:
            JobNotFound: The job is unknown to this service. Either
                it was deleted, or it never existed. Did you run()
                the job?
        """
        job = [job for job in self.list_jobs() if job.name == job_name]
        if job == []:
            raise JobNotFound()
        return job[0]

    def list_jobs(self):
        """
        List all the jobs known to the service. Note that only jobs
        that have been submitted are known to the service, if you have
        not called job.run() yet, it won't be in there.

        Returns:
            [Job]: A list of known jobs.

        Raises:
            CommunicationError: There was a problem communicating with
                the service. Is it running?
        """
        r = requests.get(self._jobs)
        if r.status_code != 200:
            raise CommunicationError(r)
        jobs_json = r.json()
        return [Job(self, job['name'], job['id'], None,
                    job['workflow'], job['input'])
                for job in jobs_json]

    def _input_dir(self, job_name):
        """
        Returns the remote URL for the input data directory for the
        given job.

        Args:
            job_name (str): The name of the job to return the
                input directory for.
        Returns:
            str: The URL.
        """
        return self._filestore + '/input/' + job_name

    def _create_input_dir(self, job_name):
        """
        Create a remote input data directory for a job on this service.

        Args:
            job_name (str): The (unique!) name of the job to make a
                directory for.

        Raises:
            JobAlreadyExists: A directory named after this job already
                exists on this service.
        """
        # Note: WebDAV requires the trailing /
        r = requests.request('MKCOL', self._input_dir(job_name) + '/')
        if r.status_code == 405:
            raise JobAlreadyExists()

    def _upload_file(self, job_name, file_desc):
        """
        Upload a local file to a remote input data directory for a
        given job. Directory must have been made first using
        _create_jobdir().

        If a file with the same name already exists, it is overwritten
        silently.

        For the file_desc argument, the following types are legal:

        str: A path to a file, whose contents and name will be used.
        Tuple[str, FileLike]: A file name to store under, and a
                file-like object to read content from, and a filename
                to store under.
        Tuple[str, bytes]: A file name to store under, and a bytes
                object with content.

        If a file-like object is passed, it will be closed after having
        been read.

        Args:
            job_name (str): The name of the job to which this file
                belongs.
            file_desc: The file's content and name, see above.

        Returns:
            str: The remote URL where the file was stored.
        """
        if isinstance(file_desc, str):
            file_name = os.path.basename(file_desc)
            try:
                content = open(file_desc, 'rb')
            except IOError as e:
                if e.errno == errno.ENOENT:
                    raise FileNotFound()
        else:
            file_name, content = file_desc
            if isinstance(content, bytes):
                content = BytesIO(content)

        try:
            remote_url = self._input_dir(job_name) + '/' + file_name
            requests.put(remote_url, data=content)
        finally:
            content.close()

        return remote_url

    def _run_job(self, job_desc):
        """
        Start a job on the service.

        Args:
            job_desc (dict): The job description. Must contain keys
                'name', 'workflow' and 'input' with the name of the
                job, the URL to the workflow, and the input description
                dict respectively.

        Returns:
            str: The id given to this job by the service.

        Raises:
            InvalidJob: The job was invalid, perhaps because the
                workflow has not been set.
        """
        r = requests.post(self._jobs, json=job_desc)
        if r.status_code == 400:
            raise InvalidJob()
        return r.json()['id']

    def _cancel_job(self, job_id):
        """
        Cancel a job running on the service.

        Args:
            job_id (str): The id of the job.

        Raises:
            JobNotFound: The job was not found on the service
            CommunicationError: There was a problem communicating with
                the service.
        """
        r = requests.post(self._jobs + '/' + job_id + '/cancel')
        if r.status_code == 404:
            raise JobNotFound()
        if r.status_code != 200:
            raise CommunicationError()


    def _job_state(self, job_id):
        """
        Get the state of the given job.

        Args:
            job_id (str): The server-side job id.

        Returns:
            str: The state that the job is in.

        Raises:
            JobNotFound: The job is unknown to the service.
            CommunicationError: There was a problem communicating with
                the service.
        """
        job = self._get_job_from_service(job_id)
        return job['state']

    def _get_job_log(self, job_id):
        """
        Get the log of the given job.

        Args:
            job_id (str): The server-side job id.

        Returns:
            str: The log as text.

        Raises:
            JobNotFound: The job is unknown to the service.
            CommunicationError: There was a problem communicating with
                the service.
        """
        r = requests.get(self._jobs + '/' + job_id + '/log')
        if r.status_code == 404:
            raise JobNotFound()
        if r.status_code != 200:
            raise CommunicationError(r)
        return r.text

    def _get_outputs(self, job_id):
        """
        Get the outputs of the given job.

        Args:
            job_id (str): The server-side job id.

        Returns:
            dict: The CWL output dictionary.

        Raises:
            JobNotFound: The job is unknown to the service.
            CommunicationError: There was a problem communicating with
                the service.
        """
        job = self._get_job_from_service(job_id)
        return job['output']

    def _delete_input_dir(self, job_name):
        """
        Delete the input directory of the job with the given name.

        Args:
            job_name (str): The client-side name of the job.

        Raises:
            JobNotFound: The input directory did not exist.
            CommunicationError: There was a problem communicating with
                the service.
        """
        import defusedxml.ElementTree as ET

        input_dir = self._input_dir(job_name)
        r = requests.request('PROPFIND', input_dir)
        xml_props = ET.fromstring(r.text)

        file_list = [path_el.text for path_el in xml_props.iter('{DAV:}href')]
        file_list.sort(key=lambda name: -len(name))

        for file_path in file_list:
            r = requests.delete(self._srv_loc + file_path)
            if r.status_code == 404:
                raise JobNotFound(r)
            if r.status_code != 204:
                raise CommunicationError(r)

    def _get_job_from_service(self, job_id):
        """
        Gets a JSON job object for the given job from the service.

        Args:
            job_id (str): The id of the job, as returned by job.id
                after it's been created.

        Returns:
            Job: The requested job.

        Raises:
            JobNotFound: The job is unknown to this service. Either
                it was deleted, or it never existed on this service.
            CommunicationError: There was an error communicating with
                the service.
        """
        r = requests.get(self._jobs + '/' + job_id)
        if r.status_code == 404:
            raise JobNotFound()
        if r.status_code != 200:
            raise error.CommunicationError(r)
        return r.json()

